[Premerge] Add flang-rt

Meinersbur opened the pull request:
Flang's runtime can now be built using LLVM's LLVM_ENABLE_RUNTIMES mechanism, with the intent to remove the old mechanism in #124126. Update the pre-merge builders to use the new mechanism.

In the current form, #124126 actually will add LLVM_ENABLE_RUNTIMES=flang-rt implicitly, so no change is strictly needed. I still think it is a good idea to do it explicitly and in advance.

I also update the dependencies:
 * flang requires mlir (but not flang-rt, that would be a circular dependency)
 * flang-rt requires flang and its transitive dependencies

On Windows, flang-rt also requires compiler-rt, but which is not building on Windows anyway.

Also consider https://github.com/llvm/llvm-zorg/pull/396

Meinersbur committed:
[Premerge] Add flang-rt

Meinersbur committed:
Don't change deps

Meinersbur committed:
flang-rt has dependencies

boomanaiden154 commented on .ci/compute-projects.sh:
```diff
@@ -47,6 +48,7 @@ function compute-projects-to-test() {
       # Flang is not stable in Windows CI at the moment
       if [[ $isForWindows == 0 ]]; then
         echo flang
+        echo flang-rt
```
Putting it here will still build it using the `LLVM_ENABLE_PROJECTS` mechanism. Is that intentional?

Meinersbur replied:
No it is not. I think I assumed that it worked like zorg's `depends_on_projects`.

But now that I look closer to it, I am even more puzzled. If there is a file changed in a runtime, let's say `libcxx`, it will not be added to `linux_projects_to_test`, and hence nothing is added to `linux_runtimes_to_test`? Am I missing something? Why does `all_projects` list also the runtimes?

boomanaiden154 replied:
They're all built as projects currently. It's on my TODO list to migrate them, but I haven't gotten around to it yet.

As far as I understand `libcxx;libcxxabi;libunwind` are the three that go through the runtimes build.

If everything works like this, I'm perfectly fine with merging and then changing it around when I switch how the runtimes build works in the premerge system. Otherwise, having it be implicit added when building flang is fine with me, and we can revisit when I switch stuff over to the proper runtimes build.

Meinersbur replied:
I added flang-rt to compute-runtimes-to-test instead. Please also review the changes in add-dependencies, I think the entry on `flang-rt` is useless then.

boomanaiden154 replied:
Have you tested this locally? I don't think this will work as is. On Linux the runtimes explicitly built through the runtimes build are in https://github.com/llvm/llvm-project/blob/6eefadd8efc0584211cb5283e0acc00a45588242/.ci/monolithic-linux.sh#L95

I would think the path to Flang would need to be passed in there? No idea how the plumbing works for `flang-rt` though. The runtimes are also built three separate times in different C++ configurations, which doesn't make sense for a library written in Fortran, although looking at the source, it seems like most of it is in C++?

If this isn't a necessary change for keeping the flang build working, it might be worth just waiting until I've had a chance to rework how we test runtimes in premerge.

Meinersbur replied:
You are correct, the path to Flang must be passed to the runtimes build. Thanks for noticing.

I have not tested it locally yet. There are some reasons for that
1. I  was assuming that the premerge-monolithic-linux builder (which I tested, see https://github.com/llvm/llvm-zorg/pull/396) was resembling the premerge configuration (which is a bootstrapping build), as was alluded to by a comment.
2. I don't have a local BuildKite infrastructure
3. I was hoping that the pre-merge check of this PR was testing it. While it was in draft mode, it did not show errors.
4. The configuration makes use of a Google-proprietary extensions, such as "-gmlt"

I will try to make it run locally anyway.

Meinersbur committed:
flang-rt is a runtime

Meinersbur committed:
Merge branch 'flang_runtime_premerge' into users/meinersbur/flang_runtime_premerge

Meinersbur committed:
Fix syntax error

Meinersbur committed:
flang_runtime_amdgpu-offload

Meinersbur committed:
Pass Flang as Fortran compiler

Meinersbur committed:
Merge branch 'main' into users/meinersbur/flang_runtime_premerge

Meinersbur committed:
Minimal changes

Meinersbur committed:
Add change in flang/ to trigger pre-merge check

Meinersbur committed:
Undo unrelated change

Meinersbur committed:
Set FLANG_ENABLE_FLANG_RT=OFF

Meinersbur commented:
I verified it locally. I also removed changes that I think are not strictly needed. So now whenever `flang` is built, `flang-rt` is built as well. `CMAKE_Fortran_COMPILER` is added to point to the Fortran compiler in the runtimes built. If flang is not being built, it is ignored by CMake.

I also added a small change to the `flang/` directory to trigger a pre-merge built of flang, to actually test it.

Some Background: In its current form #124126 will implicitly add LLVM_ENABLE_RUNTIMES=flang-rt (unless specifying `FLANG_ENABLE_FLANG_RT=OFF` that is). That is, the pre-merge will build flang-rt anyway at that point, but using the bootstrapping build and without `check-flang-rt`. For the pre-merge build this means that we get an ugly mix of bootstrapping-build of flang-rt and a runtimes-standalone build of libcxx,libcxxabi,libunwind at the same time. I am proactively adapting the buildbots to not have unfortunate surprises when #124126 is pushed, having to handle all at once / repeated reverts.

If you think that would be the lesser evil, I would be OK with skipping this PR.

Meinersbur committed:
Pass LLVM_BINARY_DIR required for testing

boomanaiden154 commented on .ci/monolithic-local.sh:
```diff
@@ -0,0 +1,127 @@
+#!/usr/bin/env bash
```
This can be dropped. `.ci/monolithic-linux.sh` works reasonably well locally, assuming you have already compute the expected projects/runtimes. Running the functions manually also isn't that difficult.

Meinersbur replied:
I added this file accidentally, it contained the modifications I needed to run. monolithic-linux.sh does not work for anyone who doesn't use Google's modified version of Clang (because of `-gmlt`).

boomanaiden154 replied:
Makes sense.

Although `-gmlt` is very much not a google-specific extension. These containers run the latest release clang built from source straight from the release branch. The option is documented in the clang options manual. https://clang.llvm.org/docs/ClangCommandLineReference.html#cmdoption-clang-gline-tables-only

boomanaiden154 reviewed (commented):
Have you tested this running premerge on `flang-rt` change? I think that would still go through the normal `LLVM_ENABLE_PROJECTS` build which might not be what you want.

If not testing changes solely to `flang-rt` is fine, then adding it to the exclude lists should be fine. If it needs to be tested in the runtimes build, adding it to the exclude list and making sure it propagates through `compute-runtimes-to-test` I think would work (after making sure `flang` is added as a dependency).

Meinersbur committed:
Adjust all runtime build configurations

Meinersbur committed:
Trigger build of Flang when flang-rt is changed

Meinersbur committed:
Test change in flang-rt to trigger flang compilation

Meinersbur committed:
Consider changes in flang-rt

Meinersbur committed:
Exclude flang-rt from LLVM_ENABLE_PROJECTS

Meinersbur commented:
> Have you tested this running premerge on `flang-rt` change? I think that would still go through the normal `LLVM_ENABLE_PROJECTS` build which might not be what you want.
> 
> If not testing changes solely to `flang-rt` is fine, then adding it to the exclude lists should be fine. If it needs to be tested in the runtimes build, adding it to the exclude list and making sure it propagates through `compute-runtimes-to-test` I think would work (after making sure `flang` is added as a dependency).

Thanks for noticing. It seems that that libcxx, etc. are already not built when they are changes (due to not being in `all_projects`), but have their own `libcxx-ci` trigger which I don't understand how it works. I first didn't intend to repair this, but seems it wasn't too difficult to add a workaround for flang-rt.

boomanaiden154 reviewed (commented):
The build failiures on the windows buildkite look relevant.

Meinersbur committed:
Do not add flang-rt to projects-to-test in the first place

Meinersbur commented:
> The build failiures on the windows buildkite look relevant.

Indeed. I should not build on Windows for something that is not even tested on Windows. 

Btw, I also noticed that it adds LLVM_ENABLE_PROJECTS=llvm, even though `llvm` is not a valid argument for `LLVM_ENABLE_PROJECTS`. I left it in the result of `compute-projects-to-test` because I think it is what triggers `check-llvm`.

boomanaiden154 commented on .ci/compute-projects.sh:
```diff
@@ -36,21 +35,30 @@ function compute-projects-to-test() {
     ;;
     *)
-      # Nothing to do
+      echo "${project}"
```
Is there a reason this moved?

boomanaiden154 reviewed (commented):
Mostly looks good to me at this point.

I'm assuming you've tested the latest version somewhere?

Meinersbur commented on .ci/compute-projects.sh:
```diff
@@ -65,6 +73,11 @@ function compute-runtimes-to-test() {
         echo $p
       done
     ;;
+    flang)
```
Not the case. `compute-runtimes-to-test` receives the aguments to "LLVM_ENABLE_PROJECTS" and prints the arguments to "LLVM_ENABLE_RUNTIMES". There is no function for the reverse direction.

If you one day rewrite the mechanism, consider giving different names to "top-level directory in the repository", "valid argument to LLVM_ENABLE_PROJECTS", and "valid argument to LLVM_ENABLE_RUNTIMES". They are currently all called "projects" in some form.

Meinersbur commented:
> I'm assuming you've tested the latest version somewhere?

Using the pre-merge itself, which seems to be more accurate than my local testing: https://buildkite.com/llvm-project/github-pull-requests/builds/154307#01956d9d-f62e-4104-b9bd-ca2c5b551686

This only tests if there is a change in the flang-rt/ directory. If you think it is necessary to also test various combinations of tests in other directories, please let me know.

boomanaiden154 commented:
> This only tests if there is a change in the flang-rt/ directory. If you think it is necessary to also test various combinations of tests in other directories, please let me know.

If you're able to test that touching something in `flang/` also works, I think that would be helpful.

Once we migrate fully over to a Github workflow I'm hoping to rewrite these scripts in Python and add some unittests which should make things a lot easier.

boomanaiden154 reviewed (approved):
Seems fine to me for now given that we're planning on reworking the runtimes build in premerge soon.

Meinersbur commented:
Tested with a flang/ change here: https://github.com/llvm/llvm-project/pull/130955. Interestingly, it build flang even on Windows: https://buildkite.com/llvm-project/github-pull-requests/builds/156355#01958a65-0b87-4bbe-bc22-6eda010c2b94

Another test with a flang/ change without any change in .ci (https://github.com/llvm/llvm-project/pull/130960) reveals that it was always that way: https://buildkite.com/llvm-project/github-pull-requests/builds/156364#01958a85-e4f9-4be5-a217-dcfcbcfed301

Reason is what I already mentioned: `compute-projects-to-test()` always echo's the directory that has changed, with this PR it just falls back to the [`*)` case](https://github.com/llvm/llvm-project/blob/b2354ffc924289ac0cbf59d53b498c121da4bd22/.ci/compute-projects.sh#L61). Still merge? I don't intend to fix all the legacy problems.

boomanaiden154 commented:
> Still merge? I don't intend to fix all the legacy problems.

Yeah, no need to fix legacy problems. We're hopefully going to rewrite this soon anyways.

Meinersbur commented:
@boomanaiden154 Thank you
